### ISSUE TITLE
Getting a type error on trying to export armature with animations

### ISSUE THREAD

**Issue by kedodrill** [KhronosGroup/glTF-Blender-IO#1485]
**Describe the bug**
I am creating an FPS and using some arms I found with a model of a crossbow I made. The [arms](https://www.blendswap.com/blend/11725) themselves will export just fine, and my crossbow does as well. However, once they are posed and I've put frames into an animation/action, I can't export it.

**To Reproduce**
1. Go to File -> Export -> gltf 2.0
2. Click on Export gltf 2.0
3. See error

**Expected behavior**
The file exports successfully as a .glb.

**Screenshots**
```
Python: Traceback (most recent call last):
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/__init__.py", line 626, in execute
    return gltf2_blender_export.save(context, export_settings)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_export.py", line 46, in save
    json, buffer = __export(export_settings)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_export.py", line 63, in __export
    __gather_gltf(exporter, export_settings)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_export.py", line 75, in __gather_gltf
    active_scene_idx, scenes, animations = gltf2_blender_gather.gather_gltf2(export_settings)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather.py", line 39, in gather_gltf2
    animations += __gather_animations(blender_scene, export_settings)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather.py", line 83, in __gather_animations
    animations_, merged_tracks = gltf2_blender_gather_animations.gather_animations(_blender_object, merged_tracks, len(animations), export_settings)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animations.py", line 76, in gather_animations
    animation = __gather_animation(blender_action, blender_object, export_settings)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animations.py", line 114, in __gather_animation
    channels=__gather_channels(blender_action, blender_object, export_settings),
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animations.py", line 149, in __gather_channels
    return gltf2_blender_gather_animation_channels.gather_animation_channels(
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_cache.py", line 65, in wrapper_cached
    result = func(*args)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_channels.py", line 74, in gather_animation_channels
    channel = __gather_animation_channel(
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_channels.py", line 202, in __gather_animation_channel
    sampler=__gather_sampler(channels, blender_object, export_settings, bake_bone, bake_channel, bake_range_start, bake_range_end, action_name, driver_obj),
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_channels.py", line 254, in __gather_sampler
    return gltf2_blender_gather_animation_samplers.gather_animation_sampler(
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_cache.py", line 65, in wrapper_cached
    result = func(*args)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_samplers.py", line 67, in gather_animation_sampler
    input=__gather_input(channels, blender_object_if_armature, non_keyed_values,
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_cache.py", line 65, in wrapper_cached
    result = func(*args)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_samplers.py", line 235, in __gather_input
    keyframes = gltf2_blender_gather_animation_sampler_keyframes.gather_keyframes(blender_object_if_armature,
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_cache.py", line 65, in wrapper_cached
    result = func(*args)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_sampler_keyframes.py", line 231, in gather_keyframes
    mat = get_bone_matrix(
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_cache.py", line 90, in wrapper_bonecache
    result = func(*args)
  File "/usr/share/blender/2.93/scripts/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_sampler_keyframes.py", line 159, in get_bone_matrix
    bpy.context.scene.frame_set(frame)
TypeError: Scene.frame_set(): error with argument 1, "frame" -  Function.frame expected an int type, not float

location: <unknown location>:-1
```

**.blend file/ .gltf**
https://1.filedit.ch/1/VlLYnIJHjXzUJNoXXIkN.blend

**Version**
 - OS: Void Linux
 - Blender Version: 2.93.5

**Additional context**
Is it possible I have an outdated version of this addon? Version shows as 1.6.16. This is more than likely me not realizing something with an animation lol, but I can't seem to find a solution :/

**Comment by julienduroure** [comment]
Yes, version of this addon for 2.93.5 is 1.6.16
I can not reproduce with your test file, export seems to be ok :
![image](https://user-images.githubusercontent.com/1696020/140409078-ffe4cb86-b984-4f92-8148-00c118138de2.png)

Can you please test on another computer and / or with a 3.0 beta version? (https://builder.blender.org/download/daily/)

Thanks!

**Comment by kedodrill** [comment]
@julienduroure Odd, it exports just fine on my laptop (Pop_OS). I downloaded 3.0 beta to my desktop (Void) and tried, possibly a different error message but it just crashes the program. When I turn off animation in the export settings, it exports successfully and no crash. I've got a Ryzen 5 1600 CPU and RX 5700 if that means anything for this. Likely something odd on my system but appreciate any ideas or help. Thank you.

```
# Python backtrace
  60   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/blender/exp/
       │ gltf2_blender_gather_animation_channels.py", line 75 in <listcomp>
  61   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/blender/exp/
       │ gltf2_blender_gather_animation_channels.py", line 75 in gather_animation_channels
  62   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/blender/exp/
       │ gltf2_blender_gather_cache.py", line 65 in wrapper_cached
  63   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/blender/exp/
       │ gltf2_blender_gather_animations.py", line 149 in __gather_channels
  64   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/blender/exp/
       │ gltf2_blender_gather_animations.py", line 114 in __gather_animation
  65   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/blender/exp/
       │ gltf2_blender_gather_animations.py", line 76 in gather_animations
  66   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/blender/exp/
       │ gltf2_blender_gather.py", line 83 in __gather_animations
  67   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/blender/exp/
       │ gltf2_blender_gather.py", line 39 in gather_gltf2
  68   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/blender/exp/
       │ gltf2_blender_export.py", line 77 in __gather_gltf
  69   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/blender/exp/
       │ gltf2_blender_export.py", line 63 in __export
  70   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/blender/exp/
       │ gltf2_blender_export.py", line 46 in save
  71   │   File "/mnt/main/downloads/blender-3.0.0-beta+v30.3e6907eb8a3d-linux.x86_64-release/3.0/scripts/addons/io_scene_gltf2/__init__.py"
       │ , line 637 in execute
```

**Comment by kedodrill** [comment]
@julienduroure alright, I've spent a little playing with this. In the script mentioned in the initial error, I've added the following below [this line](https://github.com/KhronosGroup/glTF-Blender-IO/blob/6177195cd6eca13a48a769c02aa65d63a8838183/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_sampler_keyframes.py#L157):

```python
if type(frame) is float:
    frame = int(frame)
```

and it successfully exports. I'm unsure of why this is necessary for me though. Also, I have tested this with some basic models and it is indeed an issue with armature animations - if there aren't any armatures with animations, the model exports just fine.

**Comment by scurest** [comment]
@kedodrill What is your Blender's Python version? It should say at the top of the Python Console.

**Comment by kedodrill** [comment]
> @kedodrill What is your Blender's Python version? It should say at the top of the Python Console.

3.10.0

**Comment by scurest** [comment]
Ok, that explain it.

The frame variable is indeed a float (it ultimately comes from [`fcurve.range`](https://docs.blender.org/api/current/bpy.types.FCurve.html#bpy.types.FCurve.range)), while [`scene.frame_set`](https://docs.blender.org/api/current/bpy.types.Scene.html#bpy.types.Scene.frame_set) is documented to take an integer.

If I run `C.scene.frame_set(1.0)` in my Blender (Python 3.9) I get

```
<blender_console>:1: DeprecationWarning: an integer is required (got type float).
Implicit conversion to integers using __int__ is deprecated,
and may be removed in a future version of Python.
```

So I'm guessing this has actually been removed in 3.10.

Since some distros are ahead of the curve on their Python version, this should hopefully be fixed soon (before 3.0) to avoid a repeat of #1295. The quickest fix is just to replace it with `frame_set(int(frame))` since the deprecation warning indicates that is what the current code is actually doing.

**Comment by kedodrill** [comment]
@scurest @julienduroure thank you both for the help :) I'll leave that fix in my code until 3.0 then

**Comment by chongma** [comment]
I had the same problem on Fedora 35 with default version of blender 2.93.5 with python 3.10.  i changed to flatpak version 2.93.6 and everything works now
